File: src/agentsfleetd/http/handlers/connectors/slack/status.zig
//! Slack status hook — the provider delta the generic status handler
//! (`connectors/status.zig`) dispatches to. Renders "connected" plus the
//! Slack team name when the `fleet:slack` vault handle carries a bot token,
//! else "not_connected". Never fabricates a connected state — a missing or
//! unreadable handle reads as not connected. `team` is the workspace's Slack
//! name (not a secret) so the dashboard can render "Slack connected: {team}".

const std = @import("std");
const hx_mod = @import("../../hx.zig");

const F_BOT_TOKEN = "bot_token";
const F_TEAM_NAME = "team_name";
const STATUS_CONNECTED = "connected";
const STATUS_NOT_CONNECTED = "not_connected";

/// Registry `respond_status` hook: `handle` is the parsed `fleet:slack` vault
/// object (null = missing/unreadable). Owns the full response body.
pub fn respondStatus(hx: hx_mod.Hx, handle: ?std.json.ObjectMap) void {
    const obj = handle orelse return respondNotConnected(hx);
    if (obj.get(F_BOT_TOKEN) == null) return respondNotConnected(hx);
    hx.ok(.ok, .{ .status = STATUS_CONNECTED, .team = strField(obj, F_TEAM_NAME) });
}

fn respondNotConnected(hx: hx_mod.Hx) void {
    hx.ok(.ok, .{ .status = STATUS_NOT_CONNECTED, .team = @as(?[]const u8, null) });
}

fn strField(obj: std.json.ObjectMap, key: []const u8) ?[]const u8 {
    return switch (obj.get(key) orelse return null) {
        .string => |s| s,
        else => null,
    };
}

File: rustd/crates/afd_api/src/handler/connector/status.rs
//! `GET|DELETE /v1/workspaces/{workspace_id}/connectors/{provider}` — what one
//! connection is, and letting it go.
//!
//! Two methods on one template because they are two verbs on one resource: the
//! connection either is there or is not, and both answers are about the same
//! thing. There is no PUT beside them — a connection is created by the consent
//! round-trip and cannot be asserted by writing it.
//!
//! # A disconnect revokes nothing at the provider
//!
//! It removes this daemon's sealed handle and the rows routing the provider's
//! events back, and leaves the authorization standing at the vendor. That is
//! what makes reconnecting always available after any drift — see
//! [`afd_connector::Grants::forget`], and `disconnect.zig` for the same rule
//! stated from the other side.

use std::sync::Arc;

use afd_connector::{Connection, Forgotten};
use axum::Json;
use axum::extract::{Path, State};
use axum::response::{IntoResponse as _, Response};
use http::StatusCode;
use serde::Serialize;

use super::{EVENT_READ, EVENT_WRITE, provider_of};
use crate::auth::WorkspaceContext;
use crate::handler::Refusal;
use crate::services::{Services, WorkspaceConnectors as _};

/// What a workspace holding a landed grant is told.
///
/// `oauth_status.zig`'s `STATUS_CONNECTED`, kept byte-for-byte: the dashboard
/// switches on this string and a cutover has both daemons answering the route.
const STATUS_CONNECTED: &str = "connected";

/// What a workspace holding nothing is told — see [`STATUS_CONNECTED`].
const STATUS_NOT_CONNECTED: &str = "not_connected";

/// One provider's connection, as this surface renders it.
///
/// Carries no token and no expiry. A status read answers whether a person has
/// connected and what it is called, and every other field of the stored handle
/// is the broker's business.
#[derive(Debug, Serialize)]
struct View<'v> {
    /// Whether this workspace holds a landed grant.
    status: &'static str,
    /// What a person sees the connection called, when the grant named one.
    ///
    /// Always present in the document and `null` when absent, rather than
    /// omitted: a dashboard reading `label` on an object that sometimes lacks
    /// the key would have to branch on undefined as well as on null.
    label: Option<&'v str>,
}

impl<'v> View<'v> {
    /// One connection, or the absence of one, rendered.
    fn of(connection: Option<&'v Connection>) -> Self {
        connection.map_or(
            Self {
                status: STATUS_NOT_CONNECTED,
                label: None,
            },
            |connection| Self {
                status: STATUS_CONNECTED,
                label: connection.label.as_deref(),
            },
        )
    }
}

/// `GET …/connectors/{provider}`.
///
/// Never fabricates a connected state: every shape that is not a landed grant
/// reads as not connected — see [`afd_connector::Grants::connection`].
///
/// # Errors
/// `UZ-CONN-004` for a provider this daemon does not ship, and a datastore that
/// would not answer.
pub(crate) async fn read<D: Services>(
    State(services): State<Arc<D>>,
    WorkspaceContext(owned): WorkspaceContext,
    Path((_workspace, provider_segment)): Path<(String, String)>,
) -> Result<Response, Refusal> {
    let provider = provider_of(&provider_segment)?;
    let connection = services
        .connectors()
        .connection(&owned.workspace, provider)
        .await
        .map_err(Refusal::at(EVENT_READ))?;

    Ok(Json(View::of(connection.as_ref())).into_response())
}

/// `DELETE …/connectors/{provider}`.
///
/// 204 whether or not anything was held. Idempotent in the way a delete is
/// asked to be: the caller wanted the connection gone, and it is gone —
/// answering 404 for a second press would make a person believe their first
/// one had failed.
///
/// # Errors
/// `UZ-CONN-004` for a provider this daemon does not ship, a datastore that
/// would not answer, and a vault that refused the delete.
pub(crate) async fn disconnect<D: Services>(
    State(services): State<Arc<D>>,
    WorkspaceContext(owned): WorkspaceContext,
    Path((_workspace, provider_segment)): Path<(String, String)>,
) -> Result<Response, Refusal> {
    let provider = provider_of(&provider_segment)?;
    let forgotten = services
        .connectors()
        .forget(&owned.workspace, provider)
        .await
        .map_err(Refusal::at(EVENT_WRITE))?;

    // Both outcomes answer 204, and the value is still worth matching on: a
    // reader here can see that the two were considered and deliberately
    // collapsed, where a discarded return would leave that a guess.
    match forgotten {
        Forgotten::Disconnected | Forgotten::AlreadyAbsent => {
            Ok(StatusCode::NO_CONTENT.into_response())
        }
    }
}
